### ISSUE TITLE
Pylance does not seem to identify core all() function

### ISSUE THREAD

**Issue by th3b3ginn3r** [microsoft/pyright#6595]
## Environment data
Language Server version: 2023.11.100
OS and version: MacOS 14.1.1
Python version (& distribution if applicable, e.g. Anaconda): 3.11.2

## Code Snippet

```python
def greet(_greeting: str | None, _name: str | None) -> None:
    """Greet"""

    if all([_greeting, _name]):
        caller(_greeting, _name)

    if _greeting and _name:
        caller(_greeting, _name)


def caller(_greeting: str, _name: str) -> None:
    """Caller"""
    print(f"{_greeting} {_name}")
```

## Repro Steps

1. Hover over `_greeting` or `_name` in the `all()` code block.

## Expected behavior

No Error

## Actual behavior

<img width="1189" alt="Screenshot 2023-11-30 at 3 39 25 PM" src="https://github.com/microsoft/pylance-release/assets/32792786/02511d9a-a889-47a7-93a2-e698b0947718">

```
Argument of type "str | None" cannot be assigned to parameter "_name" of type "str" in function "caller"
  Type "str | None" cannot be assigned to type "str"
    Type "None" cannot be assigned to type "str"Pylance[reportGeneralTypeIssues](https://github.com/microsoft/pyright/blob/main/docs/configuration.md#reportGeneralTypeIssues)
```

The line `if all([_greeting, _name]):`  means that both "_greeting" and "_name" must be `True` similar to the line `if _greeting and _name:`. But the there is Pylance error if using `all()` function.

**Comment by erictraut** [comment]
This is by design. It's not a bug.

Calls to the `all` function are not supported for type narrowing. For a full list of supported type narrowing patterns, refer to [this documentation](https://microsoft.github.io/pyright/#/type-concepts-advanced?id=type-guards).

Even if we were to add type narrowing support for the `all` call (which is unlikely for various reasons I've discussed elsewhere), type narrowing works only on individual symbols, not on collections of symbols. In other words, you'd need to create a list once, then assign it to a variable, and the type of that variable would be narrowed. The code above is constructing a list object using multiple symbols, then passing this object to `all`, then it constructs a completely new `list` object using the same expression.

If you want this to type check without errors, here are some options:
1. Use the `is not None` type narrowing pattern.
```python
    if _greeting is not None and _name is not None:
        caller(_greeting, _name)
```
2. Use the `x` (truthy) type narrowing pattern:
```python
    if _greeting and _name:
        caller(_greeting, _name)
```
3. Define your own custom type guard function using the mechanism defined in [PEP 647](https://peps.python.org/pep-0647/).
```python
from typing import TypeGuard, TypeVar

T = TypeVar("T")

def is_not_none(val: list[T | None]) -> TypeGuard[list[T]]:
    return all(x is not None for x in val)

def greet(_greeting: str | None, _name: str | None) -> None:
    x = [_greeting, _name]

    if is_not_none(x):
        caller(x[0], x[1])
```
